### PR TITLE
Represent unanchored microblock state

### PR DESCRIPTION
**Pull request by jcnelson**
(Don't let the size scare you -- about half the new lines are tests!)

## How It Works

This PR addresses #1617, #1700, and #1712.  In doing so, it updates the MARF and Clarity DB so that it can both represent and _persist_ state that has not been confirmed by an anchored block in a way that can be shared across threads and across DB transactions.  This is broadly useful to users and developers, because it enables the node to answer queries about the state of any _unanchored_ microblocks that have not yet been built upon by an anchored block.

The way this PR works is as follows.  At the lowest layer, it adds an alternative way of loading and storing indexed data:  _unconfirmed_ Tries.  Unlike the Tries that exist today, an unconfirmed Trie may never be committed -- it will never become part of the chain state -- but it can be stored to the underlying Trie DB as well as loaded back into RAM as a `TrieRAM` instance. 
 This is enforced by runtime checks -- a `TrieFileStorage` must be opened in "unconfirmed" mode in order to carry this out, and only an unconfirmed Trie blob row (as indicated by the DB schema) can be loaded back into RAM or dropped from the DB.

At the next layer, this PR updates the MARF so that you can either `begin()` a new Trie as before, or `begin_unconfirmed()` to create a Trie to represent unanchored microblock state off of an existing (non-unconfirmed) Trie, such that the unconfirmed Trie _will never be committed_ (enforced at runtime).  A call to `begin_unconfirmed()` will either begin a new unconfirmed Trie, or load an existing unconfirmed Trie back into RAM so that it can be subsequently updated by future calls to `insert()`.  Calling `commit()` on the MARF when opened to an unconfirmed Trie saves the unconfirmed Trie to disk so that a subsequent call to `begin_unconfirmed()` with the same `MarfTrieId` value will load it.

Each unconfirmed Trie must be instantiated off of a Trie in the canonical chain state.  To ensure that unconfirmed Tries remain "unbuildable" -- i.e. you can't begin an unconfirmed Trie on top of another unconfirmed Trie -- each unconfirmed Trie is uniquely identified by the hash of the parent Trie's hash.  This also ensures that each Trie in the chain state has a unique unconfirmed Trie that represents its anchored blocks' descendant microblock state -- it's safe to store multiple unconfirmed Tries.

The `begin_unconfirmed()` / `commit()` paradigm is expanded to the `MarfedKV` struct in the Clarity datastore, and the Clarity instance code is updated to have its own `begin_unconfirmed()` method for processing Clarity transactions that come from unanchored microblock streams.  In addition, the Clarity datastore code now has a `rollback_unconfirmed()` method that not only drops the in-RAM Trie, but also drops the unconfirmed Trie from the underlying `TrieFileStorage` and erases any metadata rows that were linked to it.  However, this method does _not_ remove any data rows that were created.  While this is a potential memory leak, it's possible to build a garbage-collector later that can remove unreferenced data rows (and, most data rows that get orphaned this way will get reclaimed once the unanchored microblock state gets confirmed, and re-applied to the database).

The updated `ClarityInstance` and `ClarityBlockConnection` are, in turn, encapsulated inside of a new struct called `UnconfirmedState` which lives inside a `StacksChainState` instance.  This struct represents a handle to the state of the chain with a particular unanchored microblock stream applied to it (namely, the stream that builds off of the canonical anchored chain tip).  The new code for this lives in `src/chainstate/stacks/db/unconfirmed.rs`.

By representing unanchored microblock state this way, it's possible for multiple threads to have their own `StacksChainState` instances while also ensuring that they will all read the same (on-disk) unconfirmed Trie data.  As long as only one thread -- the relayer thread -- is responsible for updating the unanchored state (i.e. loading the unconfirmed Trie, processing transactions, and saving it back to disk again), then all other threads will see the updates it writes and have a fresh view of the unanchored state.  This is no different from how chain state management is carried out today with anchored blocks.

The relayer thread will only track the unanchored microblock state for the canonical Stacks chain tip.  If new microblocks arrive for the canonical chain tip, and the canonical chain tip remains unchanged, then the relayer will apply the new microblocks to the canonical chain tip's unanchored microblock state.  If the canonical chain tip changes due to the arrival of one or more anchored blocks, the relayer will drop the old unanchored microblock state and reconstruct new unanchored microblock state for the new chain tip.

To make the unanchored microblock state visible to downstream clients, this PR updates some of the RPC endpoints to support a `tip=...` query parameter, which shall be a `StacksBlockId` that represents the chain tip off of which the query will run.  The affected RPC endpoints are `/v2/map_entry`, `/v2/accounts`, `/v2/contracts`.  The `tip=...` can be any block's `StacksBlockId`, or it can be the unanchored microblock state Trie's `StacksBlockId`.  The unanchored microblock state Trie's `StacksBlockId` is reported in `/v2/info` under the `unanchored_tip` key.

Maintaining unanchored microblock state over its lifetime requires _O(n^2)_ time and space for _n_ key/value pairs.  This is because the unconfirmed Trie only grows as more microblocks arrive, and the Trie must be loaded and stored to and from disk in its entirety at this time.  Fortunately, the size of a Trie is bound above by the epoch's maximum `ExecutionCost`, which limits how many leaves it can contain.

### CODE REVIEW

File: src/chainstate/stacks/mod.rs [hunk 1]
```diff
@@ -92,6 +94,7 @@ pub const STACKS_MICROBLOCK_VERSION: u8 = 0;
 pub const MAX_TRANSACTION_LEN: u32 = MAX_MESSAGE_LEN;       // TODO: shrink
 pub const MAX_BLOCK_LEN: u32 = MAX_MESSAGE_LEN;             // TODO: shrink
 
+#[derive(Serialize, Deserialize)]
```
Comment (kantai) [inline src/chainstate/stacks/mod.rs:97]: This should probably use our `impl_byte_array_serde!` macro -- that macro serializes/deserializes the byte arrays using a hex and then the serde methods for String. It makes the serialization (especially the JSON one) much friendlier.

Comment (jcnelson) [reply src/chainstate/stacks/mod.rs:97]: Will do

File: src/chainstate/stacks/db/blocks.rs [hunk 2]
```diff
@@ -2830,7 +2835,7 @@ impl StacksChainState {
                         microblocks: &Vec<StacksMicroblock>,  // parent microblocks 
                         burnchain_commit_burn: u64, 
                         burnchain_sortition_burn: u64, 
-                        user_burns: &Vec<StagingUserBurnSupport>) -> Result<(StacksHeaderInfo, Vec<StacksTransactionReceipt>), Error>
+                        user_burns: &Vec<StagingUserBurnSupport>) -> Result<(StacksHeaderInfo, Vec<StacksTransactionReceipt>, ExecutionCost, ExecutionCost), Error>
```
Comment (kantai) [inline src/chainstate/stacks/db/blocks.rs:2838]: Should this return a `StacksEpochReceipt` rather than the tuple?

Comment (jcnelson) [reply src/chainstate/stacks/db/blocks.rs:2838]: Sounds like a good idea.  Will fix.

**Comment by jcnelson** [comment]
> tests::neon_integrations::microblock_integration_test test case for querying unconfirmed state? That would give us some end to end tests for a neon miner responding to the queries.

Sure, I can do that.

**Comment by diwakergupta** [comment]
@jcnelson  Was this intended to be merged into `next`? Or perhaps I've misunderstood how we're planning to use the next branch.

**Comment by jcnelson** [comment]
It was meant to be merged to `master`.  We use `next` for incompatible changes.  Representing unanchored microblock state is a backwards-compatible change -- no existing API interfaces changed, no protocols changed, and no existing data representations changed.

**Comment by kantai** [comment]
It’s fine that this merged to ‘master’, but the ‘next’ branch needs to be kept up to date with ‘master’ so that we don’t end up generating large multi-PR conflict resolutions.

I think the most reasonable policy for this is that if you merge a PR to ‘master’ it should be the responsibility of the person who merged that PR to merge ‘master’ into ‘next’ afterwards.

**Comment by jcnelson** [comment]
> I think the most reasonable policy for this is that if you merge a PR to ‘master’ it should be the responsibility of the person who merged that PR to merge ‘master’ into ‘next’ afterwards.

Sounds good to me.  Just merged `master` to `next`.

**Pull request merged**